mount/infoParser.go

```diff
@@ -0,0 +1,1037 @@
+//
+// 2706 2192 0:155 /resolv.conf /etc/resolv.conf rw,relatime - shiftfs /var/lib/docker/containers... rw
+// 3074 2706 0:6 /null /etc/resolv.conf rw,nosuid,noexec,relatime master:2 - devtmpfs udev rw,size=4048120k,nr_inodes=1012030,mode=755
+//
+func (mi *mountInfoParser) IsOverlapMount(info *domain.MountInfo) bool {
+
+	if info == nil {
+		return false
+	}
+
+	infoParent := mi.GetParentMount(info)
+	if infoParent == nil {
+		return false
+	}
+
+	return info.MountPoint == infoParent.MountPoint
+}
+
+// IsBindMount verifies if the passed mountinfo entry is a bind-mount candidate.
+func (mi *mountInfoParser) IsBindMount(info *domain.MountInfo) bool {
+
+	if info == nil {
+		return false
+	}
+
+	mh := mi.service.mh
+	if mh == nil {
+		return false
+	}
+
+	devIdSlice := mi.devIdInfo[info.MajorMinorVer]
+
+	for _, elem := range devIdSlice {
+		if elem.MountID == info.MountID {
+			continue
+		}
+
+		if elem.Root == info.Root && elem.Source == info.Source {
+			return true
+		}
+	}
+
+	return false
+}
+
+// IsBindMount verifies if the passed mountinfo entry is a read-only bind-mount.
+func (mi *mountInfoParser) IsRoBindMount(info *domain.MountInfo) bool {
+
+	if info == nil {
+		return false
+	}
+
+	mh := mi.service.mh
+	if mh == nil {
+		return false
+	}
+
+	devIdSlice := mi.devIdInfo[info.MajorMinorVer]
+
+	for _, elem := range devIdSlice {
+		if elem.MountID == info.MountID {
+			continue
+		}
+
+		if elem.Root == info.Root && elem.Source == info.Source {
+			return mh.StringToFlags(elem.Options)&unix.MS_RDONLY == unix.MS_RDONLY
+		}
+	}
+
+	return false
+}
+
+func (mi *mountInfoParser) equivalentMounts(m1, m2 *domain.MountInfo) bool {
+
+	if m1 == nil && m2 == nil {
+		return true
+	} else if m1 == nil || m2 == nil {
+		return false
+	}
+
+	return m1.Root == m2.Root &&
+		m1.FsType == m2.FsType &&
+		m1.MajorMinorVer == m2.MajorMinorVer &&
+		m1.MountPoint == m2.MountPoint &&
+		m1.Source == m2.Source
+}
+
+func (mi *mountInfoParser) equivalentMountAncestors(m1, m2 *domain.MountInfo) bool {
+
+	if m1 == nil && m2 == nil {
+		return true
+	} else if m1 == nil || m2 == nil {
+		return false
+	}
+
+	p1 := m1.Mip.GetParentMount(m1)
+	if p1 != nil && m1.Mip.IsSysboxfsBaseMount(p1.MountPoint) {
+		p1 = m1.Mip.GetParentMount(p1)
+	}
+	p2 := m2.Mip.GetParentMount(m2)
+	if p2 != nil && m2.Mip.IsSysboxfsBaseMount(p2.MountPoint) {
+		p2 = m2.Mip.GetParentMount(p2)
+	}
+
+	return mi.equivalentMounts(p1, p2)
+}
+
+// IsCloneMount determines if the passed mountInfo entry is a 'clone' of an
+// entry in the sys-container's mount namespace.
+func (mi *mountInfoParser) IsCloneMount(
+	procInfo *domain.MountInfo,
+	ronlyMatch bool) bool {
+
+	mh := mi.service.mh
+	if mh == nil {
+		return false
+	}
+
+	var candidateList []*domain.MountInfo
+
+	// Extract the list of mountpoints matching the incoming process' maj-min-id.
+	devIdSlice := mi.devIdInfo[procInfo.MajorMinorVer]
+
+	for _, cntrInfo := range devIdSlice {
+		if cntrInfo.MountID == procInfo.MountID {
```

Comment:
Suggestion: add this comment here:
"// A mountpoint with the same ID can't be a clone (by definition)"